src/eval/lazy.rs

```diff
@@ -35,8 +35,71 @@ pub struct ThunkData {
+/// representations, but this is what we essentially do. We're just hiding the details from the
+/// user and pretend recursive records are data that can be used as a normal record, because it's
+/// both simple and natural (for the user).
+///
+/// A naive representation of recursive records could be to always keep recursive records in their
+/// functional representation, and re-compute the fixpoint at each record operation, such as field
+/// access. While correct, this would be wasteful, because the result of the fixpoint computations
+/// is actually the same throughout most record operations (put differently, `self` doesn't
+/// change). But we can't just store the computed `fixpoint` either in a normal thunk and forget
+/// about the original function, because upon merge, and more specifically upon recursive
+/// overriding happen, we do need to compute a new fixpoint.
+///
+/// Revertible thunks are a simple **memoization device** for the `fun self => ...` functional
+/// representation of a recursive record. First, note that there are several different but
+/// equivalent ways of encoding the example above as a function:
+///
+/// ```nickel
+/// fun self => { foo = self.bar + self.baz + 1, bar = 2, baz = 1 }
+///
+/// # pushing the function down the fields
+/// {foo = fun self => self.bar + 1, bar = fun self => 2, baz = fun self => 1}
+///
+/// # removing unused arguments
+/// {foo = fun self => self.bar + 1, bar = 2, baz = 1}
+///
+/// # splitting the self argument
+/// {foo = fun bar baz => bar + baz + 1, bar = 2, bar 1}
+/// ```
+///
+/// All those representations are isomorphic, in that modulo basic fee variable analysis, we can
+/// get from one to another mechanically. Pushing the functions down the fields is better because
+/// it reveals more static information about the record (such as the list of fields) without having
+/// to provide any argument first. Then, splitting `self` and removing the function altogether when
+/// the field doesn't depend on other fields is more precise with respect to dependency tracking
+/// (this has an important positive memory impact, see [ThunkDeps]).
+///
+/// In practice, Nickel uses the last representation. Fields that don't have any free variable
+/// intersecting with the name of the other fields are allocated as normal thunks that will only be
+/// evaluated at most once. Fields with actual dependencies are stored in revertible thunks, where
+/// the `orig` field stores the body of the function (in our example the expression `bar + baz +
+/// 1`), and the cached version is the application to `self` computed when we evaluated the
+/// recursive record to a normal record. The function arguments (that is, the intersection of the
+/// free variables of the expression with the set of fields) are stored inside `deps`. `deps`
+/// represent the other fields a specific field is depending on, and are guaranteed to appear free
```

Comment:
My "english senses" think it should be "freely" here as an adverb. But PL theory papers might disagree, so feel free to ignore this suggestion.
```suggestion
/// represent the other fields a specific field is depending on, and are guaranteed to appear freely
```

Reply:
Hmmmm, I actually think it's "free" here, as in "free variable" :sweat_smile: But English is not my strong suit

Reply:
I'd use "free" in this context, though as usual with English grammar I can't tell you exactly why. I'd read "are guaranteed to appear freely" as "are guaranteed to appear without restriction" which isn't really what's intended here.

Reply:
Yup, as Julien says, to appear "free" has a precise mathematical meaning, in the sense that it's a free variable of the expression. I think changing it to "freely" changes the meaning, then.